### ISSUE TITLE
New "The redirect result must start with..." error

### ISSUE THREAD

**Issue by Gitoffthelawn** [einaregilsson/Redirector#174]
Hi Einar,
While editing a bunch of Redirector rules, the latest version of Redirector (v3.5.2) is not allowing me to save my edits.
The reason is that Ridirector is now preventing rules from being saved if the redirect result does not explicitly start with "http://" or "https://".
Many of my redirects simply redirect to "$1".  Using data analysis and regex, I ensure that $1 always includes one of the allowed protocols.

Here is a common example:

> Redirect: ^http://www\.test\.com/click-.*&url=(.*)
> to: $1
> Example: http://www.test.com/click-[blob]&url=[site] → [site] 

This results in the unneeded save-blocking error:
> The redirect result must start with http:// or https://, current result is: "[site].

In this example, links to `test.com` always include either "http://" or "https://" after "url=".
I have many Redirector rules that use this template.

The easiest fix is probably just to turn the save-blocking error into a note/warning/tooltip instead of inhibiting the save operation.

(Also, as a side note, notice the extra (or missing, depending on your perspective) double-quote mark around `[site]` in the error message.)

BTW, I think it would be good in general if Redirector gracefully handles redirects to invalid URL's.  I think it currently just ignores the redirect... it might be better to redirect to a page/iframe with a little message saying something like "Redirector extension redirect destination not valid".

**Comment by einaregilsson** [comment]
But wouldn't the "fix" to this be to just put in http:// or https:// in the example url? The example url is meant to be an actual url, not something with placeholders, e.g. if you had your example url as

http://www.test.com/click-foo/?url=http://somesiteurl.com

**Comment by Gitoffthelawn** [comment]
All my Redirector example strings (and I have *many* Redirector rules) use a standard syntax to make it much more clear what the regex is actually matching and how it's working.  It's modeled after standard command syntax, as described [here](https://www.lifewire.com/how-to-read-command-syntax-2618082).

This really aids in creating, debugging, and maintaining complex rules.  It also creates a documentation of the possible characters/strings for each segment of the regular expression.

It would probably be ideal to allow each rule to have multiple examples, to test and verify each major possibility for the "redirect from" string.  This could obviate the need for versatile examples like I use.  But that would add complexity to the UI, and I know you've worked hard to keep the UI simple.  :)

**Comment by einaregilsson** [comment]
Hmm, maybe we can change it to only have it an error if a literal redirect to doesn't start with http, not if the first thing in the redirect is taken from a regex capture.

**Comment by Gitoffthelawn** [comment]
That sounds like an _**excellent**_ idea.

Let's think it out together.

It will have to allow redirects that begin with:

- $#  (where # is a positive integer)
- http$#  (where # is a positive integer)
- http://
- https://

Can you think of any others?

Although *potentially* off-topic, this might be a good time/place to bring up another hackish Redirector technique that's useful.  Take a look at this Redirector rule:

{
            "description": "Parking: Prevent redirect to Google domain parking",
            "exampleUrl": "http://domains.googlesyndication.com/apps/domainpark/",
            "exampleResult": "http://localhost/Message: This domain is parked.",
            "error": null,
            "includePattern": "^http://domains\\.googlesyndication\\.com/apps/domainpark/",
            "excludePattern": "",
            "patternDesc": "",
            "redirectUrl": "http://localhost/Message: This domain is parked.",
            "patternType": "R",
            "processMatches": "noProcessing",
            "disabled": false,
            "appliesTo": [
                "main_frame"
            ]
        }

*(Note that I've never seen this google URL called via HTTPS, so I chose - for a *slight* performance improvement - to not also match 'https://'.  If anyone finds a webpage that opens this  google URL via HTTPS, I'll update the regex to include 'https://' matching.)*

As you can see, the redirectURL goes to an intentionally non-existent page.  This allows me to simply display a message to the user via Firefox's address bar.  IIRC (and it's been several years), I previously accomplished this using a slightly different technique, but a change in either Firefox or Redirector resulting in me implementing the technique used above.

As you can also see, the above technique does not conflict at all with the current or proposed Redirector implementation.  But if you have a better way to accomplish this goal, now might be a perfect opportunity to discuss it and ensure it is considered valid by Redirector.

**Comment by einaregilsson** [comment]
Yeah, will try to get this in the next release, not sure when I'll get to
it.

As for the hackish technique, couldn't you redirect to a data url that
shows your message in the actual window? If we allowed data: urls at least.

**Comment by Gitoffthelawn** [comment]
> Yeah, will try to get this in the next release, not sure when I'll get to it. 

Great.  I look forward to it.

>As for the hackish technique, couldn't you redirect to a data url that shows your message in the actual window? If we allowed data: urls at least.

Yes, that would be ideal.  Then we could redirect to:

`data:text/plain:,[String]`

Note that in your reply, it included your private unsubscribe link.  You may want to edit it out. :)

**Comment by einaregilsson** [comment]
Thanks. I replied by email, didn't think I'd have to remove all the quoted stuff manually.

**Comment by Gitoffthelawn** [comment]
You're welcome.

Unfortunately, the GitHub web-app shows all edit history, so the link is still available when you look at the edit history of your post.  If you figure out a way to delete it without deleting the entire post, I'm interested.

**Comment by einaregilsson** [comment]
Not that worried about it, it's just to unsubscribe from this one thread as far as I know.

**Comment by Gitoffthelawn** [comment]
Yes, that's what I think/hope.  I didn't click on it to test it though!

I just don't want you to suddenly get unsubscribed from everything, or even from this thread, for that matter. :)
